### PR TITLE
frescobaldi: 2.0.16 -> 3.0.0; fix build

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

3.0.0 works with Python 3 and QT5 (proivded by `pyqt5`). These fixes are
another step towards #32883 by getting rid of the unused `poppler-qt4`.

See https://hydra.nixos.org/build/71816154/log
See ticket #36453

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ma27** [comment]
as thsi fixes the broken `frescobaldi` job on `release-18.03` as well this change should be backported and included into the upcoming release.

**Comment by garbas** [comment]
@GrahamcOfBorg build frescobaldi

**Comment by garbas** [comment]
@grahamc is 3600s limit a known problem? /me was looking at `x86_64-linux` build logs

**Comment by 7c6f434c** [comment]
It is a configuration parameter on builder machines, with 3600s being the recommended default.

File: pkgs/misc/frescobaldi/default.nix [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
+    repo = "frescobaldi";
+    rev = "v${version}";
+    sha256 = "1yn18pwsjxpxz5j3yfysmaif8k0vqahj5c7ays9cxsylpg9hl7jd";
   };
 
-  propagatedBuildInputs = with pythonPackages; [ lilypond
-    pyqt4 poppler-qt4 pygame ];
+  propagatedBuildInputs = [ lilypond pygame python-ly poppler-qt5 ];
 
-  patches = [ ./setup.cfg.patch ./python-path.patch ];
```
Comment (Ma27) [inline pkgs/misc/frescobaldi/default.nix]: if these patches aren't used anymore, they should be removed as well...

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -2366,7 +2366,7 @@ with pkgs;
 
   freetds = callPackage ../development/libraries/freetds { };
 
-  frescobaldi = callPackage ../misc/frescobaldi {};
+  frescobaldi = python3Packages.callPackage ../misc/frescobaldi {};
```
Comment (dotlambda) [inline pkgs/top-level/all-packages.nix]: Please don't use `pythonPackages.callPackage` for applications so people can override python packages used by the application.

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (Ma27) [reply pkgs/top-level/all-packages.nix]: @dotlambda I thought the same as @jtojnar, but I'm not entirely sure as I'm not a `nixpkgs` internals expert. I'll have a look at it in the next hours and comment here when I know more :)

File: pkgs/development/python-modules/poppler-qt5/default.nix [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    })
+  ];
+
+  setupPyBuildFlags = [
+    "--pyqt-sip-dir ${pyqt5}/share/sip/PyQt5"
+    "--qt-include-dir ${qtbase.dev}/include"
+  ];
+
+  nativeBuildInputs = [ pkgconfig ];
+  propagatedBuildInputs = [ sip qtbase.dev pyqt5.dev poppler ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/poppler-qt5/default.nix]: Everything that's not a python package belongs into `buildInputs`.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (dotlambda) [reply pkgs/top-level/all-packages.nix]: It's definitely possible, but you'd have to override the entire `propagatedBuildInputs`. Let's say you want to use a beta version of pyqt5, then you can just do the following when not using `pythonPackages.callPackage`:
```nix
frescobaldi.override {
  python = python.override {
    packageOverrides = self: super: {
      poppler-qt5 = super.poppler-qt5.overridePythonAttrs (oldAttrs: rec {
        version = "1.0b1";
        src = oldAttrs.src.override {
          inherit version;
          sha256 = "...";
        };
      });
    };
  };
}
```

**Comment by dotlambda** [comment]
@FRidh should definitely have a look at this.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (jtojnar) [reply pkgs/top-level/all-packages.nix]: That is not really unique to python, even if it is somewhat exaggerated by the need for package uniqueness. Perhaps there is a way to reScope a package, or apply overlay locally?

**Comment by Ma27** [comment]
@dotlambda I'm sorry if I'm misunderstanding your example, but what would be wrong with the following expression:

```
frescobaldi.override { poppler-qt5 = poppler-qt-override; }
```

I patched `fetchFromGitHub` in the example above which was part of the attribute set used to build `frescobaldi`. What I currently don't understand is why we just can't modify all other values of this attribute set, but as I said, I might be missing something %)

**Comment by FRidh** [comment]
When applications require overriding libraries, then the overriding should be done inside the expression, not in `all-packages.nix`. For deep overrides it is necessary to the package set of `python`. That's eventually why we should not use `pythonPackages.callPackage` for applications.

**Comment by Ma27** [comment]
@FRidh @jtojnar @dotlambda thanks a lot for elaborting. As two python maintainers are now in favor of injecting `pythonPackages` into the expression for building `frescobaldi`, I'll change it for now (if some folks would change their minds, we can change it later), as this is still relevant for 18.03

**Comment by dotlambda** [comment]
@GrahamcOfBorg build python2.pkgs.poppler-qt5 python3.pkgs.poppler-qt5 python2.pkgs.python-ly python3.pkgs.python-ly

**Comment by dotlambda** [comment]
@Ma27 Would you mind making 3 commits for the 3 packages involved?

**Comment by Ma27** [comment]
@dotlambda you're right, this makes the history more obvious :)

**Comment by dotlambda** [comment]
Backported in d798fa02f9c678c8b2df9987b84771e00bf20877, 29a6d29e666a365db2a804611e0bd287c5d60ba6 and 97afec420e56e38e93fd3495b68ec0aadd14b9f8.

**Pull request merged**